File: Android/FolinoReaderAndroid/src/main/kotlin/com/keynumber/folino/reader/ReaderDeviceDefaults.kt
package com.keynumber.folino.reader

import android.content.Context

/**
 * What an untouched (`null` on the Swift side) per-score Reader preference resolves to on this device.
 *
 * Keyed on `smallestScreenWidthDp` — the device's smaller dimension, which does not change when the device rotates or
 * when a window is resized. These values are what "the user never chose" resolves to, so a rule driven by the live
 * window width would re-engrave every untouched score mid-session. The `>= 600` cut is Android's own tablet
 * convention (`sw600dp`).
 *
 * The two values move together on purpose. A phone viewport is narrower than the page the score was engraved for, so
 * honoring the authored layout breaks leaves the staves cramped against an empty right margin; wrapping to the
 * viewport at a smaller staff size is what makes the same score readable.
 *
 * iOS resolves the same pair from `UIDevice.userInterfaceIdiom` (`ReaderDeviceDefaults.swift`). The numbers differ
 * between the platforms because Android engraves at a fixed layout density, so the same millimetre value renders at a
 * different apparent size.
 */
object ReaderDeviceDefaults {
    private const val TABLET_MIN_WIDTH_DP = 600

    /** Engraved staff size for a score the user has never sized themselves. */
    fun staffSize(smallestScreenWidthDp: Int): Double =
        if (smallestScreenWidthDp >= TABLET_MIN_WIDTH_DP) 24.0 else 21.0

    /** Whether a score the user has never configured reproduces the engraver's authored system / page boundaries. */
    fun honorLayoutBreaks(smallestScreenWidthDp: Int): Boolean =
        smallestScreenWidthDp >= TABLET_MIN_WIDTH_DP

    fun staffSize(context: Context): Double =
        staffSize(context.resources.configuration.smallestScreenWidthDp)

    fun honorLayoutBreaks(context: Context): Boolean =
        honorLayoutBreaks(context.resources.configuration.smallestScreenWidthDp)
}

File: Packages/Features/Reader/Sources/Reader/ReaderDeviceDefaults.swift
#if os(iOS)
import UIKit
#endif

/// What an untouched (`nil`) per-score Reader preference resolves to on this device.
///
/// Deliberately keyed on the device *idiom*, not on the live window width. These values are what `nil` resolves to, so
/// a width-driven rule would re-engrave every untouched score the moment the user rotated the device or resized a
/// Split View. The cost is that a genuinely narrow window on an iPad (Slide Over, a 1/3 Split View) still gets the
/// iPad pair — accepted: a default that is occasionally too generous beats one that moves under the reader's hands.
///
/// The two values move together on purpose. A phone viewport is narrower than the page the score was engraved for, so
/// honoring the authored `<LayoutBreak>` boundaries leaves the staves cramped against an empty right margin; wrapping
/// to the viewport at a smaller staff size is what makes the same score readable.
@MainActor
enum ReaderDeviceDefaults {
    /// Engraved staff size for a score the user has never sized themselves.
    static func staffSize(isTablet: Bool) -> Double {
        isTablet ? 14 : 12
    }

    /// Whether a score the user has never configured reproduces the engraver's authored system / page boundaries.
    static func honorLayoutBreaks(isTablet: Bool) -> Bool {
        isTablet
    }

    // The two properties below read the idiom from `UIDevice`, so they exist on iOS only. macOS has no idiom to ask:
    // `MacReaderRootScreen` calls the two portable functions above directly with `isTablet: true`, because a Mac
    // window is a large screen and wants the same generous pair an iPad gets.
    #if os(iOS)
    private static var isTablet: Bool {
        UIDevice.current.userInterfaceIdiom == .pad
    }

    static var staffSize: Double {
        staffSize(isTablet: isTablet)
    }

    static var honorLayoutBreaks: Bool {
        honorLayoutBreaks(isTablet: isTablet)
    }
    #endif
}
